src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -4,44 +4,241 @@
 
 using System;
 using System.Collections.Generic;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
 
 using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
 
 namespace ILCompiler
 {
     internal class ReadyToRunMetadataFieldLayoutAlgorithm : MetadataFieldLayoutAlgorithm
     {
         /// <summary>
-        /// CoreCLR DomainLocalModule::OffsetOfDataBlob() / sizeof(void *)
+        /// Map from EcmaModule instances to field layouts within the individual modules.
         /// </summary>
-        private const int DomainLocalModuleDataBlobOffsetAsIntPtrCount = 6;
+        private ModuleFieldLayoutMap _moduleFieldLayoutMap;
 
-        /// <summary>
-        /// CoreCLR ThreadLocalModule::OffsetOfDataBlob() / sizeof(void *)
-        /// </summary>
-        private const int ThreadLocalModuleDataBlobOffsetAsIntPtrCount = 3;
-
-        private LayoutInt _initialNonGcStaticsOffset;
+        public ReadyToRunMetadataFieldLayoutAlgorithm()
+        {
+            _moduleFieldLayoutMap = new ModuleFieldLayoutMap();
+        }
 
-        public ReadyToRunMetadataFieldLayoutAlgorithm(TargetDetails target, int numberOfTypesInModule)
+        public override ComputedStaticFieldLayout ComputeStaticFieldLayout(DefType defType, StaticLayoutKind layoutKind)
```

Comment:
Since this now overrides the whole algorithm, it should duplicate the check that prevents byref and byref-like static fields:
https://github.com/dotnet/corert/blob/c8a9707235d02e7cfc534bf2042cd4da3911b3b4/src/Common/src/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs#L210-L213

Reply:
Added.